### PR TITLE
solution 31/07

### PR DESCRIPTION
**Pull request by VadimKadasiev**

### CODE REVIEW

File: src/main/resources/db/changelog/changes/10-create-cart-items-table.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+                    foreignKeyName: shopping_cart_cart_items
+                    referencedTableName: shopping_carts
+                    referencedColumnNames: id
+              - column:
+                  name: quantity
+                  type: bigint
+                  constraints:
+                    nullable: false
+
+
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/10-create-cart-items-table.yaml]:
```suggestion
```

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            CartItem cartItem = toCartItem(cartItemRequestDto,user);
+            cartItem.setId(id);
+            cartItemRepository.save(cartItem);
+        } else {
+            throw new RuntimeException("User is not owned by this shopping cart.");
+        }
+    }
+
+    public boolean checkCartItemOwnerMatching(Long id, Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
```
Comment (okuzan) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java]: User user = (User) authentication.getPrincipal(); - make it a private method on controller level, service should deal with users

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        CartItem cartItem = cartItemRepository.findById(id).orElseThrow();
+        return cartItem.getShoppingCart().getUser().getId().equals(user.getId());
+    }
+
+    public CartItem toCartItem(CartItemRequestDto cartItemRequestDto,User user) {
```
Comment (okuzan) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
    public CartItem toCartItem(CartItemRequestDto cartItemRequestDto, User user) {
```
don't forget to Ctrl Alt L your files

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        return shoppingCartMapper.toShoppingCartResponseDto(shoppingCart);
+    }
+
+    @Override
+    public void deleteCartItem(Long id,Authentication authentication) {
+        if (checkCartItemOwnerMatching(id,authentication)) {
+            cartItemRepository.deleteById(id);
+        } else {
+            throw new RuntimeException("User is not owned by this shopping cart.");
+        }
```
Comment (okuzan) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        if (!checkCartItemOwnerMatching(id,authentication)) {
              throw new RuntimeException("User is not owned by this shopping cart.");
         }
         cartItemRepository.deleteById(id);            
      }
```

Comment (okuzan) [reply src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java]: in cases like this, invert condition in if, throw/return early, and you won't need else clause, this code is easier to read

File: src/main/java/book/store/model/ShoppingCart.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Table(name = "shopping_carts")
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = TRUE WHERE id=?")
```
Comment (okuzan) [inline src/main/java/book/store/model/ShoppingCart.java]:
```suggestion
@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = TRUE WHERE id = ?")
```

File: src/main/java/book/store/dto/CartItemRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
+    @NotNull
+    private Long bookId;
+    @NotNull
+    @Min(value = 1,message = "Quantity must be more than null!")
```
Comment (okuzan) [inline src/main/java/book/store/dto/CartItemRequestDto.java]:
```suggestion
    @Min(value = 1, message = "Quantity must be more than null!")
```

File: src/main/java/book/store/dto/CartItemRequestDto.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package book.store.dto;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
+    @NotNull
+    private Long bookId;
```
Comment (okuzan) [inline src/main/java/book/store/dto/CartItemRequestDto.java:11]: how about @ Positive ?

File: src/main/java/book/store/controller/ShoppingCartController.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    @PreAuthorize("hasRole('USER')")
+    public ShoppingCartResponseDto getShoppingCartResponseDto(Authentication authentication) {
+        return shoppingCartService.getShoppingCart(authentication);
+    }
+
+    @DeleteMapping("items/{id}")
+    @PreAuthorize("hasRole('USER')")
+    @ResponseStatus(HttpStatus.OK)
+    public void deleteCartItem(@PathVariable Long id, Authentication authentication) {
+        shoppingCartService.deleteCartItem(id,authentication);
```
Comment (okuzan) [inline src/main/java/book/store/controller/ShoppingCartController.java]:
```suggestion
        shoppingCartService.deleteCartItem(id, authentication);
```

File: src/main/java/book/store/controller/ShoppingCartController.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @PutMapping("items/{id}")
+    @PreAuthorize("hasRole('USER')")
+    @ResponseStatus(HttpStatus.OK)
+    public void updateCartItem(@PathVariable Long id,
+                               @RequestBody @Valid CartItemRequestDto cartItemRequestDto,
+                               Authentication authentication) {
+        shoppingCartService.updateCartItem(id,cartItemRequestDto,authentication);
```
Comment (okuzan) [inline src/main/java/book/store/controller/ShoppingCartController.java]:
```suggestion
        shoppingCartService.updateCartItem(id, cartItemRequestDto,authentication);
```

**Review by okuzan — CHANGES_REQUESTED** [review]
Good job overall!

File: src/main/java/book/store/controller/ShoppingCartController.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    @PutMapping("items/{id}")
+    @PreAuthorize("hasRole('USER')")
+    @ResponseStatus(HttpStatus.OK)
+    public void updateCartItem(@PathVariable Long id,
+                               @RequestBody @Valid CartItemRequestDto cartItemRequestDto,
+                               Authentication authentication) {
+        shoppingCartService.updateCartItem(id, cartItemRequestDto, getCurrentUser(authentication));
+    }
+
+    private User getCurrentUser(Authentication authentication) {
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/ShoppingCartController.java]:
```suggestion
    private Long getCurrentUser(Authentication authentication) {
```

provide userId straightaway

File: src/main/java/book/store/controller/ShoppingCartController.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    @PreAuthorize("hasRole('USER')")
+    public void createCartItem(
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartResponseDto createCartItem(
```

File: src/main/java/book/store/controller/ShoppingCartController.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    @PreAuthorize("hasRole('USER')")
+    @ResponseStatus(HttpStatus.OK)
+    public void deleteCartItem(@PathVariable Long id, Authentication authentication) {
+        shoppingCartService.deleteCartItem(id, getCurrentUser(authentication));
+    }
+
+    @PutMapping("items/{id}")
+    @PreAuthorize("hasRole('USER')")
+    @ResponseStatus(HttpStatus.OK)
+    public void updateCartItem(@PathVariable Long id,
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartResponseDto updateCartItem(@PathVariable Long id,
```

File: src/main/java/book/store/dto/CartItemRequestDto.java [hunk 13]
```diff
@@ -0,0 +1,12 @@
+package book.store.dto;
+
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
+    @Positive
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/dto/CartItemRequestDto.java:9]:
```suggestion
    @Positive
    @NotNull
```

File: src/main/java/book/store/dto/CartItemRequestDto.java [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+    private Long bookId;
+    @Positive(message = "Quantity must be more than null!")
+    private Long quantity;
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/dto/CartItemRequestDto.java]:
```suggestion
    private int quantity;
```

File: src/main/java/book/store/dto/CartItemRequestDto.java [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package book.store.dto;
+
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
+    @Positive
+    private Long bookId;
+    @Positive(message = "Quantity must be more than null!")
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/dto/CartItemRequestDto.java]: Wrong message, Positive isn't about null

File: src/main/java/book/store/model/CartItem.java [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+@Entity
+@Getter
+@Setter
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "shopping_cart_id", nullable = false)
```

File: src/main/java/book/store/model/CartItem.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+    private ShoppingCart shoppingCart;
+
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/CartItem.java:28]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "book_id", nullable = false)
```

File: src/main/java/book/store/model/CartItem.java [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+    private Book book;
+
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/CartItem.java]:
```suggestion
```

File: src/main/java/book/store/model/CartItem.java [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    private Long quantity;
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/CartItem.java]:
```suggestion
    private int quantity;
```

File: src/main/java/book/store/model/ShoppingCart.java [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+@SQLRestriction(value = "is_deleted = FALSE")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @OneToOne
+    private User user;
+
+    @OneToMany(mappedBy = "shoppingCart")
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/ShoppingCart.java]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/book/store/model/ShoppingCart.java [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/ShoppingCart.java]:
```suggestion
     @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id", nullable = false)
```

File: src/main/java/book/store/repository/ShoppingCartRepository.java [hunk 22]
```diff
@@ -0,0 +1,12 @@
...
+
+import book.store.model.ShoppingCart;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.repository.query.Param;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    @EntityGraph(attributePaths = "cartItems")
+    Optional<ShoppingCart> findShoppingCartByUserId(@Param("userId") Long userId);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/repository/ShoppingCartRepository.java]: Where this `@Param` will be used?

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+        ShoppingCart shoppingCart = shoppingCartRepository
+                .findShoppingCartByUserId(user.getId()).orElseThrow();
+        return shoppingCartMapper.toShoppingCartResponseDto(shoppingCart);
+    }
+
+    @Override
+    public void deleteCartItem(Long id, User user) {
+        if (!checkCartItemOwnerMatching(id, user)) {
+            throw new RuntimeException("User is not owned by this shopping cart.");
+        }
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart shoppingCart = repository.findByUserId(userId);
        CartItem cartItem = itemRepository.findByIdAndShoppingCartId(itemId, shoppingCart.getId())
                .orElseThrow(.....
```

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+    @Override
+    public void createCartItem(CartItemRequestDto cartItemRequestDto, User user) {
+        CartItem cartItem = toCartItem(cartItemRequestDto, user);
+        cartItemRepository.save(cartItem);
+    }
+
+    @Override
+    public ShoppingCartResponseDto getShoppingCart(User user) {
+        ShoppingCart shoppingCart = shoppingCartRepository
+                .findShoppingCartByUserId(user.getId()).orElseThrow();
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java]: .orElseThrow(); ????

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+    @Override
+    public void createShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
+    public void createCartItem(CartItemRequestDto cartItemRequestDto, User user) {
+        CartItem cartItem = toCartItem(cartItemRequestDto, user);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java]: Book book = bookRepository.findById
ShoppingCart cart = repository.findByUserId(userId)
cart.getCartItems().stream()......

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    @Override
+    public void updateCartItem(Long id, CartItemRequestDto cartItemRequestDto, User user) {
+        if (!checkCartItemOwnerMatching(id, user)) {
+            throw new RuntimeException("User is not owned by this shopping cart.");
+        }
+        CartItem cartItem = toCartItem(cartItemRequestDto, user);
+        cartItem.setId(id);
+        cartItemRepository.save(cartItem);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart cart = repository.findByUserId...
        CartItem cartItem = itemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId())
        
        update quantity
        save(cartItem)
        shopCartMapper....
```

File: src/main/java/book/store/service/impl/UserServiceImpl.java [hunk 27]
```diff
@@ -33,6 +36,8 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         Role userRole = roleRepository.findByName(Role.RoleName.ROLE_USER);
         user.setRoles(Set.of(userRole));
         User savedUser = userRepository.save(user);
+        shoppingCartService.createShoppingCart(userRepository
+                .findByEmail(user.getEmail()).orElseThrow());
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/UserServiceImpl.java]:
```suggestion
        userRepository.save(user);
        shoppingCartService.createShoppingCart(user);
```

File: src/main/resources/db/changelog/changes/10-create-cart-items-table.yaml [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+                  name: shopping_cart_id
+                  type: bigint
+                  constraints:
+                    nullable: false
+                    foreignKeyName: shopping_cart_cart_items
+                    referencedTableName: shopping_carts
+                    referencedColumnNames: id
+              - column:
+                  name: quantity
+                  type: bigint
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/10-create-cart-items-table.yaml]:
```suggestion
                  type: int
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/book/store/controller/ShoppingCartController.java [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+    public ShoppingCartResponseDto createCartItem(
+            @RequestBody @Valid CartItemRequestDto cartItemRequestDto,
+            Authentication authentication) {
+        return shoppingCartService.createCartItem(cartItemRequestDto,
+                getCurrentUserId(authentication));
+    }
+
+    @GetMapping
+    @PreAuthorize("hasRole('USER')")
+    public ShoppingCartResponseDto getShoppingCartResponseDto(Authentication authentication) {
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartResponseDto getShoppingCart(Authentication authentication) {
```

File: src/main/java/book/store/controller/ShoppingCartController.java [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+        return shoppingCartService.getShoppingCart(getCurrentUserId(authentication));
+    }
+
+    @DeleteMapping("items/{id}")
+    @PreAuthorize("hasRole('USER')")
+    @ResponseStatus(HttpStatus.OK)
+    public ShoppingCartResponseDto deleteCartItem(@PathVariable Long id,
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/ShoppingCartController.java]:
```suggestion
    public void deleteCartItem(@PathVariable Long id,
```

File: src/main/java/book/store/dto/CartItemRequestDto.java [hunk 31]
```diff
@@ -0,0 +1,15 @@
...
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
+    @Positive
+    @NotNull
+    private Long bookId;
+    @Positive(message = "Quantity must be more than 0!")
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/dto/CartItemRequestDto.java]:
```suggestion
```

int can't be null

File: src/main/java/book/store/repository/CategoryRepository.java [hunk 32]
```diff
@@ -5,4 +5,5 @@
 import org.springframework.stereotype.Repository;
 
 @Repository
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/repository/CategoryRepository.java]: If you use this annotation (that is not necessary), use it for all repositories, be consistent

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 33]
```diff
@@ -0,0 +1,99 @@
...
+    private final ShoppingCartMapper shoppingCartMapper;
+
+    @Override
+    public void createShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java:35]: Should we use Transactional?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java (hunk 33 above)
Comment (VadimKadasiev) [reply src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java:35]: OK. But i wonder, how should it be inconsistent, if there's only one saving to repository action?

Comment (Elena-Bruyako) [reply src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java:35]: hmm..I think there is misunderstanding. 

I texted about consistency related to `@Repository `annotation

File: src/main/java/book/store/security/FieldMatch.java [hunk 34]
```diff
@@ -9,8 +9,12 @@
 @Constraint(validatedBy = FieldMatchValidator.class)
 public @interface FieldMatch {
     String message();
+
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/security/FieldMatch.java]: Replace to `validation` package

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
one comment

**Review by okuzan — APPROVED** [review]
good job

**Pull request merged**